hs_file_types/models/reftimeseries.py

```diff
@@ -0,0 +1,922 @@
+        return ['WOF', "WPS", "DirectFile"]
+
+    @classmethod
+    def get_allowed_return_types(cls):
+        """returns a list of returnType controlled vocabulary terms"""
+        return ['WaterML 1.1', 'WaterML 2.0', 'TimeseriesML']
+
+    @classmethod
+    def get_allowed_service_types(cls):
+        """returns a list of serviceType controlled vocabulary terms"""
+        return ['SOAP', 'REST']
+
+    @classmethod
+    def create(cls):
+        # this custom method MUST be used to create an instance of this class
+        rf_ts_metadata = RefTimeseriesFileMetaData.objects.create(json_file_content="No data")
+        return cls.objects.create(metadata=rf_ts_metadata)
+
+    @classmethod
+    def set_file_type(cls, resource, file_id, user):
+        """
+            Sets a json resource file to RefTimeseriesFile type
+            :param resource: an instance of resource type CompositeResource
+            :param file_id: id of the resource file to be set as RefTimeSeriesFile type
+            :param user: user who is setting the file type
+            :return:
+            """
+
+        log = logging.getLogger()
+
+        # get the the selected resource file object
+        res_file = utils.get_resource_file_by_id(resource, file_id)
+
+        if res_file is None:
+            raise ValidationError("File not found.")
+
+        if res_file.extension != '.refts':
+            raise ValidationError("Not a Ref Time Series file.")
+
+        files_to_add_to_resource = []
+        if res_file.has_generic_logical_file:
+            try:
+                json_file_content = _validate_json_file(res_file)
+            except Exception as ex:
+                raise ValidationError(ex.message)
+
+            # get the file from irods to temp dir
+            temp_file = utils.get_file_from_irods(res_file)
+            temp_dir = os.path.dirname(temp_file)
+            files_to_add_to_resource.append(temp_file)
+            file_folder = res_file.file_folder
+            with transaction.atomic():
+                # first delete the json file that we retrieved from irods
+                # for setting it to reftimeseries file type
+                delete_resource_file(resource.short_id, res_file.id, user)
+
+                # create a reftiemseries logical file object to be associated with
+                # resource files
+                logical_file = cls.create()
+
+                logical_file.metadata.json_file_content = json_file_content
+                logical_file.metadata.save()
+
+                try:
+                    # add the json file back to the resource
+                    uploaded_file = UploadedFile(file=open(temp_file, 'rb'),
+                                                 name=os.path.basename(temp_file))
+
+                    new_res_file = utils.add_file_to_resource(
+                        resource, uploaded_file, folder=file_folder
+                    )
```

Comment:
@hyi I agree. We should certainly look at this at some point. However, I am not sure how we can avoid this file transfer issue in the case where we are updating a resource file due to metadata changes. For example, netcdf file and SQLite file. May be python based irods microservice can solve this issue.